common/metrics/service_wrapper.go

```diff
@@ -0,0 +1,272 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"sync"
+	"time"
+
+	"github.com/uber-go/tally"
+	"github.com/uber/tchannel-go/thrift"
+	m "go.uber.org/cadence/.gen/go/cadence"
+	"go.uber.org/cadence/.gen/go/shared"
+)
+
+type workflowServiceMetricsWrapper struct {
+	service     m.TChanWorkflowService
+	scope       tally.Scope
+	childScopes map[string]tally.Scope
+	mutex       sync.Mutex
+}
+
+// NewWorkflowServiceWrapper creates a new wrapper to WorkflowService that will emit metrics for each service call.
+func NewWorkflowServiceWrapper(service m.TChanWorkflowService, scope tally.Scope) m.TChanWorkflowService {
+	return &workflowServiceMetricsWrapper{service: service, scope: scope, childScopes: make(map[string]tally.Scope)}
+}
+
+func (w *workflowServiceMetricsWrapper) getScope(scopeName string) tally.Scope {
+	w.mutex.Lock()
+	scope, ok := w.childScopes[scopeName]
+	if ok {
+		w.mutex.Unlock()
+		return scope
+	}
+	scope = w.scope.SubScope(scopeName)
+	w.childScopes[scopeName] = scope
+	w.mutex.Unlock()
+	return scope
+}
+
+func (w *workflowServiceMetricsWrapper) DeprecateDomain(ctx thrift.Context, request *shared.DeprecateDomainRequest) error {
+	scope := w.getScope("DeprecateDomain")
+	scope.Counter(CadenceRequest).Inc(1)
+	startTime := time.Now()
+	err := w.service.DeprecateDomain(ctx, request)
+	scope.Timer(CadenceLatency).Record(time.Now().Sub(startTime))
+	if err != nil {
```

Comment:
We should not count error if the error is EntityNotExistsError, that is a valid response.